src/common/mfu_io.c

```diff
@@ -25,81 +25,136 @@
 static int mpi_rank;
 
 #ifdef DAOS_SUPPORT
-struct d_hash_table *dir_hash;
-
-struct mfu_dir_hdl {
-        d_list_t	entry;
-        dfs_obj_t	*oh;
-        char		name[PATH_MAX];
+/* Handle for a hash table entry */
+struct daos_dir_hdl {
+    d_list_t    entry;
+    dfs_obj_t*  oh;
+    char*       name;
 };
 
-static inline struct mfu_dir_hdl* hdl_obj(d_list_t *rlink)
+/* Return a newly allocated daos_dir_hdl structure */
+static struct daos_dir_hdl* daos_dir_hdl_new(void)
 {
-        return container_of(rlink, struct mfu_dir_hdl, entry);
+    struct daos_dir_hdl* hdl = (struct daos_dir_hdl*) MFU_MALLOC(sizeof(struct daos_dir_hdl));
+    hdl->oh = NULL;
+    hdl->name = NULL;
+
+    return hdl;
 }
 
-static bool key_cmp(struct d_hash_table *htable, d_list_t *rlink,
-	const void *key, unsigned int ksize)
+/* free a daos_dir_hdl structure */
+static void daos_dir_hdl_delete(struct daos_dir_hdl** phdl)
 {
-        struct mfu_dir_hdl *hdl = hdl_obj(rlink);
+    if (phdl != NULL) {
+        struct daos_dir_hdl* hdl = *phdl;
+        if (hdl->oh != NULL) {
+            dfs_release(hdl->oh);
+        }
+        mfu_free(&hdl->name);
+        mfu_free(&hdl);
```

Comment:
Let's pass the caller's argument here rather than the address of the local variable.
```suggestion
        mfu_free(phdl);
```
The difference is that this will set the caller's pointer to NULL.

Reply:
Ah, yes. Thank you!